system/lib/wasmfs/syscalls.cpp

```diff
@@ -253,12 +253,32 @@ __wasi_errno_t __wasi_fd_pread(__wasi_fd_t fd,
 __wasi_errno_t __wasi_fd_close(__wasi_fd_t fd) {
   auto fileTable = wasmFS.getLockedFileTable();
 
+  if (!fileTable[fd]) {
+    return __WASI_ERRNO_BADF;
+  }
+
   // Remove openFileState entry from fileTable.
   fileTable[fd] = nullptr;
 
   return __WASI_ERRNO_SUCCESS;
 }
 
+__wasi_errno_t __wasi_fd_sync(__wasi_fd_t fd) {
+  auto openFile = wasmFS.getLockedFileTable()[fd];
+  if (!openFile) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  // Nothing to flush for anything but a data file, but also not an error either
+  // way.
```

Comment:
I don't think this will be correct once we have more nontrivial backends and caching layers that may not immediately write directory data to backing stores. Looking at the docs for `__wasi_fd_sync`, they mention that it is supposed to be similar to `fsync`, and the docs for `fsync` talk about calling `fsync` on directory file descriptors as something you might want to do, so we should support that.

Reply:
Yes, I can add a TODO here.